crates/re_renderer/src/allocator/data_texture_source.rs

```diff
@@ -0,0 +1,373 @@
+        self.buffers.iter().map(|b| b.capacity()).sum()
+    }
+
+    /// The number of elements that can be written without allocating more memory.
+    #[inline]
+    pub fn remaining_capacity(&self) -> usize {
+        self.buffers
+            .iter()
+            .skip(self.active_buffer_index)
+            .map(|b| b.remaining_capacity())
+            .sum()
+    }
+
+    /// Ensure invariant that the active buffer has some remaining capacity or is the next buffer that needs to be allocated.
+    #[inline]
+    fn ensure_active_buffer_invariant(&mut self) {
+        if self.active_buffer_index < self.len()
+            && self.buffers[self.active_buffer_index].remaining_capacity() == 0
+        {
+            self.active_buffer_index += 1;
+        }
+
+        // Note that if we're *very* unlucky there might be a lot of unused buffers.
+        // This happens only if there's quadratic growing `reserve` calls without any writes.
+        debug_assert!(self.buffers.len() >= self.active_buffer_index);
+        // If the active buffer exists, it must have remaining capacity.
+        debug_assert!(
+            self.buffers.len() == self.active_buffer_index
+                || self.buffers[self.active_buffer_index].remaining_capacity() > 0
+        );
+        // The buffer before the active buffer must be full.
+        debug_assert!(
+            self.active_buffer_index == 0
+                || self.buffers[self.active_buffer_index - 1].remaining_capacity() == 0
+        );
+    }
+
+    /// Ensures that there's space internally for at least `num_elements` more elements.
+    ///
+    /// Creating new buffers is a relatively expensive operation, so it's best to
+    /// reserve gratuitously and often. Ideally, you know exactly how many elements you're going to write and reserve
+    /// accordingly at the start.
+    ///
+    /// If there's no more space, a new buffer is allocated such that:
+    /// * have a total capacity for at least as many elements as requested
+    /// * be at least double the size of the last buffer
+    /// * keep it easy to copy to textures by always being a multiple of the maximum row size we use for data textures
+    ///      -> this massively simplifies the buffer->texture copy logic!
+    pub fn reserve(&mut self, num_elements: usize) -> Result<(), CpuWriteGpuReadError> {
+        let remaining_capacity = self.remaining_capacity();
+        if remaining_capacity >= num_elements {
+            return Ok(());
+        }
+
+        let last_buffer_size = self.buffers.last().map_or(0, |b| b.capacity());
+        let new_buffer_size = (num_elements - remaining_capacity)
+            .max(last_buffer_size * 2)
+            .next_multiple_of(self.max_data_texture_width());
+
+        self.buffers
+            .push(self.ctx.cpu_write_gpu_read_belt.lock().allocate(
+                &self.ctx.device,
+                &self.ctx.gpu_resources.buffers,
+                new_buffer_size,
+            )?);
+
```

Comment:
According to the docs of `active_buffer_index`: prior to `reserve`, `active_buffer_index` can point to a full buffer, but after it may not. So you need to call `ensure_active_buffer_invariant` here.

Reply:
ah. docs are wrong now. The commit before changed around where and exactly what invariant is enforced but failed to update the docs. Previously it was at the end of reserve calls, but I made the mistake of not doing so unconditionally. I figured since `reserve` calls are both internal and external it makes it needlessly hard to enforce validity and perf wise this should make a difference

The new wording for the invariant is now a bit more precise (350931632ae495edf511618629f9e205553037df):

```
    /// Buffer in which data is currently written.
    ///
    /// Between all public calls:
    /// * all buffers before are full
    /// * all buffers after are empty (if any)
    /// * the buffer at this index either does not exist or has remaining capacity
```